# Integration of multiple scRNA-seq datasets from different labs
With the accumulation of single cell RNA (scRNA-seq) studies comes the burning question: how can we compare datasets generated in completely distinct ways?

This question is fundamental because we want to be able to integrate datasets produced by different labs that report on similar biological specimens. As we used two different technical approaches in [Tabula muris](https://github.com/czbiohub/tabula-muris-vignettes/blob/master/README.md), our dataset therefore offers the ideal platform for techincal comparisons of sequencing results derived from identically isolated biological samples but processed with two distinct methods for scRNA-seq
<!--
<ol>
  <li><span style="font-weight: bold;">microfluidic droplet-based 3â€™-end counting</span> that enabled the survey of thousands of cells at relatively low coverage;</li>
  <li><span style="font-weight: bold;">FACS-based full length transcript analysis</span> that enabled characterization of cell types with high sensitivity and coverage.</li>
</ol>
--->

<!-- Our dataset therefore offers the ideal platform for techincal comparisons of sequencing results derived from identically isolated biological samples but processed with two distinct methods for scRNA-seq [technical comparisons](linktothepaperfigures).
-->

<!--We illustrate the variability to be expected between methods and experiments by comparing our two measurement approaches to one another, and to data from Han et al (Cell 2018) generated using a third method, microwell-seq. One striking feature is the variability in the number of genes detected per cell between organs and tissues and between methods. -->

<!--[figure]-->

<!--This difference does not appear to be due to sequencing depth, as the microfluidic droplet libraries are nearly saturated (Supp. Fig. 4b) and deeper sequencing of the FACS libraries could only increase the number of genes detected.-->

<!--[figure]-->

<!--We also looked at how the number of detected genes across each organ changes with different thresholds on the number of reads or UMIs (Supp. Fig. 5). -->

<!--[figure]-->

<!--We found that the number of detected genes decreases monotonically with increasing thresholds at similar rates across different organs and tissues within each method. We observed that in the droplet data more than half of the detected genes are represented by only a single UMI; this is to be expected given that only a few thousand UMIs are captured per cell. The FACS data are sampled much more deeply and one needs to set a relatively high threshold of 40 reads to see a comparable reduction in gene detection sensitivity.-->

In parallel, [Han et al](http://www.cell.com/cell/abstract/S0092-8674(18)30116-8) developed a new method, microwell-seq, and also derived another mouse cell atlas. It is pertinent to compare these two works because together they offer a more complete picture of mouse biology at single cell resolution. 
Traditionally we compare the data at the level of differentially expressed genes, to avoid entering the realm of batch effects introduced by different technologies. To do so, we computed [lists of genes](https://github.com/czbiohub/tabula-muris/blob/master/22_markers/FindMarkers.Rmd) that differentiate each cell type from the rest of the cells for a given organ across all three methods, focusing on common organs and cell types for the three methods. The following code snippets are from the Matlab scripts available [here](https://github.com/czbiohub/tabula-muris/blob/master/22_markers/methods_comparison.m) and for simplicity  we also provide the [gene signature files](https://github.com/czbiohub/tabula-muris/tree/master/22_markers) for each organ. There are 8 tissues for which we have data from the three methods, so we can get the signatures for the tissue specific cell ontologies for all methods:

```Matlab
tissuesAvailable = {'Bladder', 'Kidney', 'Liver', 'Lung','Marrow', 'Limb_Muscle', 'Spleen', 'Thymus'}';
```
<!--
[s,v] = listdlg('PromptString','Select a tissue:',...
    'SelectionMode','single',...
    'ListString',tissuesAvailable);


tissuename = tissuesAvailable{s};filename1 = sprintf('facs_%s_cell_ontology_class_markers.csv',tissuename);
tissueFACS = readtable(filename1);
filename11 = sprintf('facs_%s_cell_ontology_class_classes.csv',tissuename);
tissueFACSnames = readtable(filename11, 'delimiter', ',');
filename2 = sprintf('droplet_%s_cell_ontology_class_markers.csv',tissuename);
tissue10x = readtable(filename2);
filename22 = sprintf('droplet_%s_cell_ontology_class_classes.csv',tissuename);
tissue10xnames = readtable(filename22, 'delimiter', ',');

if strcmp(tissuename,'Limb_Muscle')
	tissuenameHan = 'Muscle';
else tissuenameHan = tissuename;
end
filename3 = sprintf('%s_Han.csv',tissuenameHan);
tissueMicrowellSeq = readtable(filename3);

```

Next we get the cell signatures for all methods
-->

We start by getting the tissue specific cell ontologies signatures for Tabula muris FACS:
```Matlab
cellNamesTMfacs = unique(table2cell(tissueFACSnames(:,2)));
cellNamesTMfacs = sortrows(cellNamesTMfacs);
for i = 1:length(cellNamesTMfacs)
	[a,b] = ismember(table2cell(tissueFACSnames(:,2)),cellNamesTMfacs(i));
	cellClustersTMfacs(i,1:length(find(a))) = str2double(table2cell(tissueFACSnames(a,1)))';
end
cellClustersTMfacs = cellClustersTMfacs - ones;

tissueTMfacs = table2cell(tissueFACS);
for i = 1:size(cellClustersTMfacs,1)
	tissueTMfacsCellSigsAUX = [];
	tissueTMfacsCellSigsFCaux = [];
    for j = 1:size(cellClustersTMfacs,2)
    	[m,n] = find(str2double(tissueTMfacs(:,7))==cellClustersTMfacs(i,j));

    	tissueTMfacsCellSigsAUX = [tissueTMfacsCellSigsAUX;tissueTMfacs(m,8)];
    	tissueTMfacsCellSigsFCaux = [tissueTMfacsCellSigsFCaux;tissueTMfacs(m,3)];
    end
	[tissueTMfacsCellSigsAUX,ia,ic] = unique(tissueTMfacsCellSigsAUX);
	tissueTMfacsCellSigs(1:length(tissueTMfacsCellSigsAUX)+1,i) = [cellNamesTMfacs(i);tissueTMfacsCellSigsAUX];
	tissueTMfacsCellSigsFC(1:length(tissueTMfacsCellSigsAUX)+1,i) = [-cellClustersTMfacs(i,j);tissueTMfacsCellSigsFCaux(ia)];
end
```

Then for Tabula muris droplet:
```Matlab
cellNamesTM10x = unique(table2cell(tissue10xnames(:,2)));
cellNamesTM10x = sortrows(cellNamesTM10x);
for i = 1:length(cellNamesTM10x)
	[a,b] = ismember(table2cell(tissue10xnames(:,2)),cellNamesTM10x(i));
    	cellClustersTM10x(i,1:length(find(a))) = str2double(table2cell(tissue10xnames(a,1)))';
end
cellClustersTM10x = cellClustersTM10x - ones;

tissueTM10x = table2cell(tissue10x);
for i = 1:size(cellClustersTM10x,1)
	tissueTM10xCellSigsAUX = [];
	tissueTM10xCellSigsFCaux = [];
    for j = 1:size(cellClustersTM10x,2)
    	[m,n] = find(str2double(tissueTM10x(:,7))==cellClustersTM10x(i,j));

    	tissueTM10xCellSigsAUX = [tissueTM10xCellSigsAUX;tissueTM10x(m,8)];
    	tissueTM10xCellSigsFCaux = [tissueTM10xCellSigsFCaux;tissueTM10x(m,3)];
    end
	[tissueTM10xCellSigsAUX,ia,ic] = unique(tissueTM10xCellSigsAUX);
	tissueTM10xCellSigs(1:length(tissueTM10xCellSigsAUX)+1,i) = [cellNamesTM10x(i);tissueTM10xCellSigsAUX];
	tissueTM10xCellSigsFC(1:length(tissueTM10xCellSigsAUX)+1,i) = [-cellClustersTM10x(i,j);tissueTM10xCellSigsFCaux(ia)];
end
```

And finally for microwell-seq:
```Matlab
microwellSeq = table2cell(tissueMicrowellSeq);
microwellSeqcelltypes = {};
aux = 1;
microwellSeqCellSigs = cell(length(microwellSeq),length(unique(microwellSeq(:,8))));
for i = 1:length(microwellSeq)
    if isequal(microwellSeq(i,8),{[]})
    	microwellSeq(i,8) = cellaux;
    	microwellSeq{i,9} = aux;
    else
    	cellaux = microwellSeq(i,8);
    	aux = aux+1;
    	microwellSeq{i,9} = aux;
    	microwellSeqcelltypes = [microwellSeqcelltypes;cellaux];
        if i > 1
        	microwellSeqCellSigs(1:(i-baux+1),length(microwellSeqcelltypes)-1) = [microwellSeq(baux,8);microwellSeq(baux:i-1,6)];
        	microwellSeqcellSigsFC(1:(i-baux+1),length(microwellSeqcelltypes)-1) = [-cell2mat(microwellSeq(baux,5));cell2mat(microwellSeq(baux:i-1,2))];
        end
    	baux = i;
    end
end
microwellSeqCellSigs(1:(i-baux+1),end) = [microwellSeq(baux,8);microwellSeq(baux:i-1,6)];
microwellSeqcellSigsFC(1:(i-baux+1),end+1) = [-cell2mat(microwellSeq(baux,5));cell2mat(microwellSeq(baux:i-1,2))];

[microwellSeqcelltypesSorted,index] = sortrows(microwellSeqcelltypes);
microwellSeqCellSigs = microwellSeqCellSigs(:,index);
microwellSeqcellSigsFC = microwellSeqcellSigsFC(:,index);
```

Once this is set we search for the genes that overlap across the three methods and plot it as Venn diagrams.

<!--
Before we are able to do so, we have to manually match the cell ontologies, defining  compare the technologies.

if strcmp(tissuename,'Bladder')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;2] [1;2] [[8 11 12];[14 0 0]]];
end

if strcmp(tissuename , 'Kidney')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;2;3;4],[1;5;3;6],[[8 0 0 0 0 0 0];[12 13 17 18 19 20 21];[5 10 11 0 0 0 0];[14 15 0 0 0 0 0]]];
end

if strcmp(tissuename,'Liver')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[3;4;1],[2;3;4],...
    	[[5 0 0 0];[11 12 16 17];[1 2 19 20]]];
end

if strcmp(tissuename, 'Lung')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;3;4;8;10;11;12],[1;3;5;8;10;11;13],[[6 22 0 0 0 0 0 0 0 0 0];[15 32 0 0 0 0 0 0 0 0 0];[8 0 0 0 0 0 0 0 0 0 0];...
    	[18 19 20 0 0 0 0 0 0 0 0];[10 11 12 13 14 17 21 23 24 26 28];[25 0 0 0 0 0 0 0 0 0 0];[29 30 31 0 0 0 0 0 0 0 0]]];
end

if strcmp(tissuename,'Marrow')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[8;14;17],[8;11;12],[[5 0];[6 7];[9 10]]];
end

if strcmp(tissuename,'Limb_Muscle')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;2;3;4;5;6],[1;3;4;5;6;7],[[1 2 0];[17 0 0];[4 0 0];[8 9 0];[16 0 0];[10 11 12]]];
end

if strcmp(tissuename , 'Spleen')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;2;3],[1;2;4],[[6];[11];[5]]];
end

if strcmp(tissuename , 'Thymus')
	tissueOverlaps = cell(1,6);
	tissueOverlapsSizes = [];
	overlapMatrix = [[1;2],[1;2],[[3 0];[4 2]]];
end

```
-->

<!--
This data struct has the following structure:
<ul>
  <li>1st column: cell types for the organ being compared as defined in the cell ontology associated with the FACS dataset</li>
  <li>2nd column: corresponding cell types as defined in the cell ontology associated with the 10x dataset</li>
  <li>3rd:end columns: corresponding cell types as defined in the cell ontology associated with the microwellSeq dataset</li>
</ul

Lastly we computed the overlaps and drew the Venn diagrams>
-->

```Matlab
% find the overlaps
overlapGenesTissue = {};
for i = 1:size(overlapMatrix,1)

        % FACS
    	aux1 = tissueTMfacsCellSigs(~cellfun('isempty',tissueTMfacsCellSigs(:,overlapMatrix(i,1))),overlapMatrix(i,1));
    	aux = aux1(2:end);
        % 10x
    	baux1 = tissueTM10xCellSigs(~cellfun('isempty',tissueTM10xCellSigs(:,overlapMatrix(i,2))),overlapMatrix(i,2));
    	baux = baux1(2:end);
        % microwellSeq
    	caux = [];
        for j = 3:size(overlapMatrix,2)
            if overlapMatrix(i,j)
            	cauxaux = microwellSeqCellSigs(~cellfun('isempty',microwellSeqCellSigs(:,overlapMatrix(i,j))),overlapMatrix(i,j));
            	cauxaux = cauxaux(2:end);
            	caux = [caux;cauxaux];
            end
        end
    	caux = unique(caux);

        % check overlaps for FACS
    	[a1,b1] = ismember(aux,baux);
    	overlap_FACS_10x = aux(a1);
    	[a2,b2] = ismember(overlap_FACS_10x,caux);
    	overlap_FACS_10x_microwell = overlap_FACS_10x(a2);
    	overlap_FACS_10x_nomicrowell = overlap_FACS_10x(~a2);

    	overlap_FACSno10x = aux(~a1);
    	[a3,b3] = ismember(overlap_FACSno10x,caux);
    	overlap_FACSno10x_microwell = overlap_FACSno10x(a3);
    	overlap_FACSno10xnomicrowell = overlap_FACSno10x(~a3);

        % check overlaps for 10x
    	[c1,d1] = ismember(baux,aux);
    	overlap_10x_FACS = baux(c1);
    	[c2,d2] = ismember(overlap_10x_FACS,caux);
    	overlap_10x_FACS_microwell = overlap_10x_FACS(c2); % should be the same as overlap_FACS_10x_microwell
    	overlap_10x_FACS_nomicrowell = overlap_10x_FACS(~c2); % should be the same as overlap_FACS_10xnomicrowell

    	overlap_10xnoFACS = baux(~c1);
    	[c3,d3] = ismember(overlap_10xnoFACS,caux);
    	overlap_10xnoFACS_microwell = overlap_10xnoFACS(c3);
    	overlap_10xnoFACSnomicrowell = overlap_10xnoFACS(~c3);

        % check overlaps for microwellSeq
    	[e1,f1] = ismember(caux,aux);
    	overlap_microwell_FACS = caux(e1);
    	[e2,f2] = ismember(overlap_microwell_FACS,baux);
    	overlap_microwell_FACS_10x = overlap_microwell_FACS(e2); % should be the same as overlap_FACS_10x_microwell
    	overlap_microwell_FACSno10x = overlap_microwell_FACS(~e2); % should be the same as overlap_FACSno10x_microwell

    	overlap_microwellnoFACS = caux(~e1);
    	[e3,f3] = ismember(overlap_microwellnoFACS,baux);
    	overlap_microwellnoFACS_10x = overlap_microwellnoFACS(e3); % should be the same as overlap_10xnoFACS_microwell
    	overlap_microwellnoFACSno10x = overlap_microwellnoFACS(~e3);

    	tissueOverlaps{1,i} = {overlap_FACS_10x_microwell; overlap_FACS_10x_nomicrowell; overlap_FACSno10x_microwell; overlap_FACSno10xnomicrowell; ...
        	overlap_10xnoFACS_microwell; overlap_10xnoFACSnomicrowell; overlap_microwellnoFACSno10x};

    	tissueOverlapsSizes = [tissueOverlapsSizes [length(overlap_FACS_10x_microwell);length(overlap_FACS_10x_nomicrowell);length(overlap_FACSno10x_microwell);length(overlap_FACSno10xnomicrowell);...
        	length(overlap_10xnoFACS_microwell);length(overlap_10xnoFACSnomicrowell);length(overlap_microwellnoFACSno10x)]];

        % plot Venn diagrams
    	vennNumbers = [length(overlap_FACS_10x_microwell);length(overlap_FACS_10x_nomicrowell);length(overlap_FACSno10x_microwell);length(overlap_FACSno10xnomicrowell);...
        	length(overlap_10xnoFACS_microwell);length(overlap_10xnoFACSnomicrowell);length(overlap_microwellnoFACSno10x)];
    	vennNumbers = vennNumbers([4,2,6,5,7,3,1]);
    	eRR = vennX(vennNumbers,.1);
    	title([sprintf('Tissue: %s',tissuename) sprintf(' -- CellType: %s',cellNamesTMfacs{overlapMatrix(i,1)})])
    	saveas(gcf,[sprintf('Tissue_%s',tissuename) sprintf('_CellType_%s',cellNamesTMfacs{overlapMatrix(i,1)})],'pdf')       

        % write gene overlaps
    	overlapGenes = cell(max([length(overlap_FACS_10x_microwell);length(overlap_FACS_10x_nomicrowell);length(overlap_FACSno10x_microwell);length(overlap_FACSno10xnomicrowell);...
        	length(overlap_10xnoFACS_microwell);length(overlap_10xnoFACSnomicrowell);length(overlap_microwellnoFACSno10x)]),7);

    	overlapGenesAux = {overlap_FACSno10xnomicrowell overlap_10xnoFACSnomicrowell overlap_microwellnoFACSno10x...
        	overlap_FACS_10x_nomicrowell overlap_FACSno10x_microwell overlap_10xnoFACS_microwell...
        	overlap_FACS_10x_microwell};

        for k = 1:length(overlapGenesAux)
        	overlapGenes(1:length(overlapGenesAux{k}),k) = overlapGenesAux{k};
        end

    	overlapGenes = [{'FACS','Droplet10x','microwellSeq','FACS_10x','FACS_microwellSeq','microwellSeq_10x','FACS_10x_microwellSeq'};overlapGenes];
    	[oga,ogb] = size(overlapGenes);
    	aux = cell(oga,2);
    	aux(1:2,2) = ['Cell Ontology';cellNamesTMfacs(overlapMatrix(i,1))];
        if ~isempty(overlapGenesTissue)
        	[ogta,ogtb] = size(overlapGenesTissue);
            if oga > ogta
            	overlapGenesTissue = [overlapGenesTissue;cell(oga-ogta,ogtb)];
            	overlapGenesTissue = [overlapGenesTissue aux overlapGenes];
            else
            	overlapGenesTissue(1:oga,end+1:end+2+ogb) = [aux overlapGenes];
            end
        else overlapGenesTissue = [aux overlapGenes];
        end

end


% write to table overlapping genes in the Venn diagrams
overlapGenesTissue = cell2table(overlapGenesTissue);
writetable(overlapGenesTissue,sprintf('%s_genes.csv',tissuename))

% write to table cell ontology matches
overlapMatrixNames = cell(size(overlapMatrix));
overlapMatrixNames(:,1) = cellNamesTMfacs(overlapMatrix(:,1));
overlapMatrixNames(:,2) = cellNamesTM10x(overlapMatrix(:,2));
for i=1:size(overlapMatrix,1)
    for j = 3:size(overlapMatrix,2)
        if overlapMatrix(i,j)
        	overlapMatrixNames(i,j) = microwellSeqcelltypesSorted(overlapMatrix(i,j));
        end
    end
end

microwellSeqVars = {'microwellSeq1','microwellSeq2','microwellSeq3','microwellSeq4',...
    'microwellSeq5','microwellSeq6','microwellSeq7','microwellSeq8','microwellSeq9',...
    'microwellSeq10','microwellSeq11','microwellSeq12','microwellSeq13','microwellSeq14'};
overlapMatrixNames = cell2table(overlapMatrixNames,'VariableNames',{'FACS','Droplet10x',microwellSeqVars{1:size(overlapMatrix,2)-2}});
writetable(overlapMatrixNames,sprintf('%s_cell_ontologies.csv',tissuename))

```

The graphical output is [here](https://github.com/czbiohub/tabula-muris-vignettes/blob/master/vignettes/methods_comparison/methods_comparison_diffExp.pdf).

For this example we manually set the ```overlapMatrix```, that is, we forced the algorithm to compare cell ontologies that we matched by ontology name. The problem with doing so is that we are assuming that the annotations are consistent across all experiments but this might not necessarily be the case. Let's take the current example - while in Tabula muris we used [OBO Cell Ontology]( http://obofoundry.org/ontology/cl.html) structured and controlled vocabulary for cell types in animals, Han et al in the microwell seq paper took a dramaitcally different approach and have cell types named as ```Neutrophil_Npg_high``` and ```Neutrophil_Retngg_high```, for example.


<embed src="https://github.com/czbiohub/tabula-muris-vignettes/blob/master/vignettes/methods_comparison/methods_comparison_diffExp.pdf" width="500" height="375" type='application/pdf'>

In this example we have illustrated one way of comparing datasets that have been independently obtained without the need to re-analize the raw data, instead we take the fully analysed datasets and mix&match across experiments. As a challenge why don't you try to compare all cells from the same cell ontology across all organs, as for example ```macrophages```?

In the next post we will be looking at using machine learning techniques to derive a (probably) better, more scalable comparisons pipeline.

&#35;tabula-muris &#35;sc-RNA-seq &#35;methods-comparison &#35;FACS &#35;10x-genomics &#35;droplets-based-methods &#35;microwell-seq
